### PR TITLE
Speed up computing h3 cell bounding box with great circle edges

### PR DESCRIPTION
**Pull request by iverase**
In https://github.com/elastic/elasticsearch/pull/92460, we changed the way to compute the bounding box of an h3 cell for geo_points to use spherical geometry, that is considering edges as great circles. In order to do that we compute the bounding box using lucene spatial3d library.

This works but it is pretty wasteful as the spatial3d needs to transform our latitudes and longitudes to the x/y/z representation and then transform the back to lat / lon. In addition this library allocates many objects so it is not best suitable when called in tight loops.

I researched a bit and found this interesting [paper](https://trs.jpl.nasa.gov/bitstream/handle/2014/41271/07-0286.pdf) where it describes a way to compute the bounding box for great circle edges using the [Clairaut's formula.](https://streckenflug.at/download/formeln.pdf)

I implemented it and it gives a very nice speed up so let use this one instead:

```
|                                       Total Young Gen GC time |            |    2.733       |    0.626       |    -2.107   |      s |  -77.09% |
|                                      Total Young Gen GC count |            | 1672           |  391           | -1281       |        |  -76.61% |
|                                         Total Old Gen GC time |            |    0           |    0           |     0       |      s |    0.00% |
|                                        Total Old Gen GC count |            |    0           |    0           |     0       |        |    0.00% |
|                                                Min Throughput | geohexgrid |    0.184475    |    0.255139    |     0.07066 |  ops/s |  +38.31% |
|                                               Mean Throughput | geohexgrid |    0.186729    |    0.259933    |     0.0732  |  ops/s |  +39.20% |
|                                             Median Throughput | geohexgrid |    0.187095    |    0.260596    |     0.0735  |  ops/s |  +39.29% |
|                                                Max Throughput | geohexgrid |    0.187547    |    0.261286    |     0.07374 |  ops/s |  +39.32% |
|                                       50th percentile latency | geohexgrid | 5269.69        | 3790.82        | -1478.87    |     ms |  -28.06% |
|                                       90th percentile latency | geohexgrid | 5458.49        | 3849.12        | -1609.37    |     ms |  -29.48% |
|                                       99th percentile latency | geohexgrid | 5691.31        | 4112.37        | -1578.94    |     ms |  -27.74% |
|                                      100th percentile latency | geohexgrid | 5778.38        | 4319.43        | -1458.95    |     ms |  -25.25% |
|                                  50th percentile service time | geohexgrid | 5269.69        | 3790.82        | -1478.87    |     ms |  -28.06% |
|                                  90th percentile service time | geohexgrid | 5458.49        | 3849.12        | -1609.37    |     ms |  -29.48% |
|                                  99th percentile service time | geohexgrid | 5691.31        | 4112.37        | -1578.94    |     ms |  -27.74% |
|                                 100th percentile service time | geohexgrid | 5778.38        | 4319.43        | -1458.95    |     ms |  -25.25% |
|                                                    error rate | geohexgrid |    0           |    0           |     0       |      % |    0.00% |
```

### CODE REVIEW

File: libs/h3/src/main/java/org/elasticsearch/h3/LatLng.java [hunk 1]
```diff
@@ -22,9 +22,14 @@
  */
 package org.elasticsearch.h3;
 
+import java.util.Objects;
+
 /** pair of latitude/longitude */
 public final class LatLng {
 
+    /** Minimum Angular resolution. */
+    public static final double MINIMUM_ANGULAR_RESOLUTION = Math.PI * 1.0e-12; // taken from lucene's spatial3d
```
Comment (craigtaverner) [inline libs/h3/src/main/java/org/elasticsearch/h3/LatLng.java:31]: Very high resolution. If I'm right, this is about 1% of a millimeter.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtil.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        }
+        if (minLat < 0) {
+            // we only correct the min latitude if negative
+            minLat = boundary.getLatLon(minLatPos).greatCircleMinLatitude(computeEdge(boundary, minLatPos, MIN_COMPARATOR));
+        }
+        if (maxLat > 0) {
+            // we only correct the max latitude if positive
+            maxLat = boundary.getLatLon(maxLatPos).greatCircleMaxLatitude(computeEdge(boundary, maxLatPos, MAX_COMPARATOR));
+        }
+        // the min / max longitude is computed the same way as in cartesian, being careful with polygons crossing the dateline
```
Comment (craigtaverner) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtil.java:107]: I guess, as with cartesian, we do not allow geometries wider that 180, except for the polar cells which are 360 degrees. I presume we've verified that only polar cells are wider than 180?

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtilTests.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.elasticsearch.test.ESTestCase;
+
+public class H3SphericalUtilTests extends ESTestCase {
+
+    private static final double DELTA = 1e-8;
+
+    public void testRandomBounds() {
+        GeoBoundingBox boundingBox = new GeoBoundingBox(new GeoPoint(), new GeoPoint());
+        for (int res = 1; res <= H3.MAX_H3_RES; res++) {
+            final long h3 = H3.geoToH3(90, 0, res);
```
Comment (craigtaverner) [inline x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtilTests.java:24]: The test name implies this test should use randomly chosen h3, but it seems to only use the north pole. Perhaps you meant to generate a random h3?

**Review by craigtaverner — CHANGES_REQUESTED** [review]
This looks great, but there is one test that I think was supposed to use a randomly generated h3, but always uses the north pole.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtil.java (hunk 2 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtil.java:107]: Yes, that's correct.

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtilTests.java (hunk 3 above)
Comment (iverase) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtilTests.java]: Yes, I added randomization. 

Curiously I had to adjust the delta for the bounding box longitude. It seems spatial3d might adjust slightly those values. Still I think should not be a problem for this other way to compute the bounding box.

**Review by craigtaverner — APPROVED** [review]
LGTM

**Pull request merged**